pkg/controller/cachemanager/cachemanager_test.go

```diff
@@ -0,0 +1,455 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+)
+
+var cfg *rest.Config
+
+func TestMain(m *testing.M) {
+	testutils.StartControlPlane(m, &cfg, 3)
+}
+
+// TestCacheManager_AddObject_RemoveObject tests that we can add/ remove objects in the cache.
+func TestCacheManager_AddObject_RemoveObject(t *testing.T) {
+	cm, _, ctx := makeCacheManagerForTest(t, false, false)
+
+	pod := fakes.Pod(
+		fakes.WithNamespace("test-ns"),
+		fakes.WithName("test-name"),
+	)
+	unstructuredPod, err := runtime.DefaultUnstructuredConverter.ToUnstructured(pod)
+	require.NoError(t, err)
+
+	require.NoError(t, cm.AddObject(ctx, &unstructured.Unstructured{Object: unstructuredPod}))
+
+	// test that pod is cache managed
+	opaClient, ok := cm.opa.(*fakes.FakeOpa)
+	require.True(t, ok)
+	require.True(t, opaClient.HasGVK(pod.GroupVersionKind()))
+
+	// now remove the object and verify it's removed
+	require.NoError(t, cm.RemoveObject(ctx, &unstructured.Unstructured{Object: unstructuredPod}))
+	require.False(t, opaClient.HasGVK(pod.GroupVersionKind()))
+}
+
+// TestCacheManager_processExclusion makes sure that we don't add objects that are process excluded.
+func TestCacheManager_processExclusion(t *testing.T) {
+	cm, _, ctx := makeCacheManagerForTest(t, false, false)
+	processExcluder := process.Get()
+	processExcluder.Add([]configv1alpha1.MatchEntry{
+		{
+			ExcludedNamespaces: []util.Wildcard{"test-ns-excluded"},
+			Processes:          []string{"sync"},
+		},
+	})
+	cm.processExcluder.Replace(processExcluder)
+
+	pod := fakes.Pod(
+		fakes.WithNamespace("test-ns-excluded"),
+		fakes.WithName("test-name"),
+	)
+	unstructuredPod, err := runtime.DefaultUnstructuredConverter.ToUnstructured(pod)
+	require.NoError(t, err)
+	require.NoError(t, cm.AddObject(ctx, &unstructured.Unstructured{Object: unstructuredPod}))
+
+	// test that pod from excluded namespace is not cache managed
+	opaClient, ok := cm.opa.(*fakes.FakeOpa)
+	require.True(t, ok)
+	require.False(t, opaClient.HasGVK(pod.GroupVersionKind()))
+}
+
+// TestCacheManager_errors tests that the cache manager responds to errors from the opa client.
+func TestCacheManager_errors(t *testing.T) {
+	cm, _, ctx := makeCacheManagerForTest(t, false, false)
+	opaClient, ok := cm.opa.(*fakes.FakeOpa)
+	require.True(t, ok)
+	opaClient.SetErroring(true) // This will cause AddObject, RemoveObject to err
+
+	pod := fakes.Pod(
+		fakes.WithNamespace("test-ns"),
+		fakes.WithName("test-name"),
+	)
+	unstructuredPod, err := runtime.DefaultUnstructuredConverter.ToUnstructured(pod)
+	require.NoError(t, err)
+
+	// test that cm bubbles up the errors
+	require.ErrorContains(t, cm.AddObject(ctx, &unstructured.Unstructured{Object: unstructuredPod}), "test error")
+	require.ErrorContains(t, cm.RemoveObject(ctx, &unstructured.Unstructured{Object: unstructuredPod}), "test error")
+}
+
+// TestCacheManager_listAndSyncData tests that the cache manager can add gvks to the data store.
```

Comment:
This test is either doing more than it says it does or than it means to do.  I'm not sure.

If we want to test that data gets added to the store, then we could:

1. Set up the fake API server with fake objects.
2. Configure CM to sync those objects.
3. Confirm those objects are in the cache.

This test also includes a manual call to wipe and then adds more gvks and then checks again.  If we want to check other specific pieces of functionality like:

1. Does a wipe yield an empty cache?
2. Do two updates to the SyncAggregator get aggregated correctly in a list call?

It could then be helpful to write shorter, more directed tests for these ideas.  That way, if we break part of the functionality, the test that points to specifically what is broken will fail.

If combining it all is what we're going for, I find that's often better achieved in an e2e test.  As long as it is stable, it provides a very strong signal (aka if it is working, it's definitely working).  This test does parts of the flow (like wiping) manually, and thus could continue working even if we broke part of the functionality.

Reply:
Added unit tests for the scenarios described above.